File: Vanilla/sources/zomboid/zombie/inventory/types/Key.java
package zombie.inventory.types;

import java.io.IOException;
import java.nio.ByteBuffer;
import zombie.inventory.InventoryItem;
import zombie.inventory.ItemType;
import zombie.scripting.objects.Item;

public final class Key extends InventoryItem {
   private int keyId = -1;
   private boolean padlock = false;
   private int numberOfKey = 0;
   private boolean digitalPadlock = false;
   public static final Key[] highlightDoor = new Key[4];

   public Key(String var1, String var2, String var3, String var4) {
      super(var1, var2, var3, var4);
      this.cat = ItemType.Key;
   }

   public int getSaveType() {
      return Item.Type.Key.ordinal();
   }

   public void takeKeyId() {
      if (this.getContainer() != null && this.getContainer().getSourceGrid() != null && this.getContainer().getSourceGrid().getBuilding() != null && this.getContainer().getSourceGrid().getBuilding().def != null) {
         this.setKeyId(this.getContainer().getSourceGrid().getBuilding().def.getKeyId());
      }

   }

   public static void setHighlightDoors(int var0, InventoryItem var1) {
      if (var1 instanceof Key && !((Key)var1).isPadlock() && !((Key)var1).isDigitalPadlock()) {
         highlightDoor[var0] = (Key)var1;
      } else {
         highlightDoor[var0] = null;
      }

   }

   public int getKeyId() {
      return this.keyId;
   }

   public void setKeyId(int var1) {
      this.keyId = var1;
   }

   public String getCategory() {
      return this.mainCategory != null ? this.mainCategory : "Key";
   }

   public void save(ByteBuffer var1, boolean var2) throws IOException {
      super.save(var1, var2);
      var1.putInt(this.getKeyId());
      var1.put((byte)this.numberOfKey);
   }

   public void load(ByteBuffer var1, int var2) throws IOException {
      super.load(var1, var2);
      this.setKeyId(var1.getInt());
      this.numberOfKey = var1.get();
   }

   public boolean isPadlock() {
      return this.padlock;
   }

   public void setPadlock(boolean var1) {
      this.padlock = var1;
   }

   public int getNumberOfKey() {
      return this.numberOfKey;
   }

   public void setNumberOfKey(int var1) {
      this.numberOfKey = var1;
   }

   public boolean isDigitalPadlock() {
      return this.digitalPadlock;
   }

   public void setDigitalPadlock(boolean var1) {
      this.digitalPadlock = var1;
   }
}

File: Vanilla/sources/zdoc/media/lua/shared/Library/Key.lua
---@class Key : zombie.inventory.types.Key
---@field private keyId int
---@field private padlock boolean
---@field private numberOfKey int
---@field private digitalPadlock boolean
---@field public highlightDoor Key[]
Key = {}

---throws java.io.IOException
---
---Overrides:
---
---save in class InventoryItem
---@public
---@param output ByteBuffer
---@param net boolean
---@return void
function Key:save(output, net) end

---@public
---@param numberOfKey int
---@return void
function Key:setNumberOfKey(numberOfKey) end

---@public
---@param arg0 ByteBuffer
---@param arg1 int
---@return void
function Key:load(arg0, arg1) end

---@public
---@return boolean
function Key:isDigitalPadlock() end

---Get the key number of the building and set it to the key
---@public
---@return void
function Key:takeKeyId() end

---@public
---@param digitalPadlock boolean
---@return void
function Key:setDigitalPadlock(digitalPadlock) end

---Overrides:
---
---getCategory in class InventoryItem
---@public
---@return String
function Key:getCategory() end

---@public
---@return int
function Key:getSaveType() end

---@public
---@param arg0 int
---@param arg1 InventoryItem
---@return void
function Key:setHighlightDoors(arg0, arg1) end

---@public
---@param padlock boolean
---@return void
function Key:setPadlock(padlock) end

---Overrides:
---
---setKeyId in class InventoryItem
---@public
---@param keyId int
---@return void
function Key:setKeyId(keyId) end

---@public
---@return boolean
function Key:isPadlock() end

---Overrides:
---
---getKeyId in class InventoryItem
---@public
---@return int
function Key:getKeyId() end

---@public
---@return int
function Key:getNumberOfKey() end
